transaction.go

```diff
@@ -49,198 +90,303 @@ type Transaction struct {
+			t.initComplete <- build.ExtendErr("Couldn't write the page to file", err)
+			return
 		}
-		data = append(data, bytes...)
 		page = page.nextPage
 	}
-	return crypto.HashBytes(data), nil
 }
 
 // validateChecksum checks if a transaction has been corrupted by computing a hash
-// and comparing it to the one in the finalPage of the transaction
+// and comparing it to the one in the firstPage of the transaction
 func (t Transaction) validateChecksum() error {
 	// Check if finalPage is set
-	if t.finalPage == nil {
-		return errors.New("Couldn't verify checksum. finalPage is nil")
+	if t.firstPage == nil {
+		return errors.New("Couldn't verify checksum. firstPage is nil")
 	}
 	checksum, err := t.checksum()
 	if err != nil {
 		return errors.New("Failed to create checksum for validation")
 	}
-	if checksum != t.finalPage.transactionChecksum {
+	if checksum != t.firstPage.transactionChecksum {
 		return errors.New("checksum not valid")
 	}
 	return nil
 }
 
-//commit commits a transaction by setting the correct status and checksum
-func (t *Transaction) commit() error {
-	// set the status of the first page first
-	t.firstPage.pageStatus = pageStatusComitted
-
-	// calculate the checksum and write it to the last page
-	checksum, err := t.checksum()
-	if err != nil {
-		return build.ExtendErr("Unable to create checksum of transaction", err)
-	}
-	t.finalPage.transactionChecksum = checksum
-	err = t.finalPage.Write(t.wal.logFile)
-	if err != nil {
-		return build.ExtendErr("Writing the final page failed", err)
+// NewTransaction creates a transaction from a set of updates
+func (w *WAL) NewTransaction(updates []Update) *Transaction {
+	// Create new transaction
+	newTransaction := Transaction{
+		Updates:      updates,
+		wal:          w,
+		initComplete: make(chan error),
 	}
 
-	// Finalize the commit by writing the first page with the updated status
-	err = t.firstPage.Write(t.wal.logFile)
-	if err != nil {
-		return build.ExtendErr("Writing the first page failed", err)
-	}
+	// Initialize the transaction by splitting up the payload among free pages
+	// and writing them to disk.
+	go threadedInitTransaction(&newTransaction)
 
-	t.commitComplete = true
-	t.wal.logFile.Sync()
-	return nil
+	return &newTransaction
 }
 
-// SignalApplyComplete  informs the WAL that it is safe to free the used pages to reuse them in a new transaction
-func (t *Transaction) SignalApplyComplete() <-chan error {
+// SignalUpdatesApplied  informs the WAL that it is safe to free the used pages to reuse them in a new transaction
+func (t *Transaction) SignalUpdatesApplied() <-chan error {
 	if !t.setupComplete || !t.commitComplete || t.releaseComplete {
-		panic("misuse of transaction - call each of the signaling methods exactly ones, in serial, in order")
+		panic("misuse of transaction - call each of the signaling methods exactly once, in serial, in order")
 	}
 	t.releaseComplete = true
 	notifyChannel := make(chan error)
 
 	go func() {
```

Comment:
Can probably split this goroutine into it's own function. Need to have `notifyChannel` as an input, but I think that's it.

Reply:
actually, if we change the function to be blocking instead of using a goroutine, we don't need to split this code out.